packages/contracts/contracts/extensions/DutchAuction/DutchAuction.sol

```diff
@@ -0,0 +1,150 @@
+    /// @return matchedFillResults amounts filled and fees paid by maker and taker of matched orders.
+    function matchOrders(
+        LibOrder.Order memory buyOrder,
+        LibOrder.Order memory sellOrder,
+        bytes memory buySignature,
+        bytes memory sellSignature
+    )
+        public
+        returns (LibFillResults.MatchedFillResults memory matchedFillResults)
+    {
+        AuctionDetails memory auctionDetails = getAuctionDetails(sellOrder);
+        // Ensure the auction has not yet started
+        require(auctionDetails.currentTimeSeconds >= auctionDetails.beginTimeSeconds, "AUCTION_NOT_STARTED");
+        // Ensure the auction has not expired. This will fail later in 0x but we can save gas by failing early
+        require(sellOrder.expirationTimeSeconds > auctionDetails.currentTimeSeconds, "AUCTION_EXPIRED");
+        // Ensure the auction goes from high to low
+        require(auctionDetails.beginAmount > auctionDetails.endAmount, "INVALID_AMOUNT");
+        // Validate the buyer amount is greater than the current auction amount
+        require(buyOrder.makerAssetAmount >= auctionDetails.currentAmount, "INVALID_AMOUNT");
+        // Match orders, maximally filling `buyOrder`
+        matchedFillResults = EXCHANGE.matchOrders(
+            buyOrder,
+            sellOrder,
+            buySignature,
+            sellSignature
+        );
+        // Return any spread to the seller
+        uint256 leftMakerAssetSpreadAmount = matchedFillResults.leftMakerAssetSpreadAmount;
+        if (leftMakerAssetSpreadAmount > 0) {
+            // Calculate the excess from the buy order. This can occur if the buyer sends in a higher
+            // amount than the calculated current amount
+            uint256 buyerExcessAmount = buyOrder.makerAssetAmount-auctionDetails.currentAmount;
+            uint256 sellerExcessAmount = leftMakerAssetSpreadAmount-buyerExcessAmount;
+            bytes memory assetData = sellOrder.takerAssetData;
+            address token = assetData.readAddress(16);
+            if (sellerExcessAmount > 0) {
+                address makerAddress = sellOrder.makerAddress;
+                IERC20Token(token).transfer(makerAddress, sellerExcessAmount);
+            }
+            if (buyerExcessAmount > 0) {
+                address takerAddress = buyOrder.makerAddress;
+                IERC20Token(token).transfer(takerAddress, buyerExcessAmount);
+            }
+        }
+        return matchedFillResults;
+    }
+
+    /// @dev Calculates the Auction Details for the given order
+    /// @param order The sell order
+    /// @return AuctionDetails
+    function getAuctionDetails(
+        LibOrder.Order memory order
+    )
+        public
+        returns (AuctionDetails memory auctionDetails)
+    {
+        uint256 makerAssetDataLength = order.makerAssetData.length;
+        // We assume auctionBeginTimeSeconds and auctionBeginAmount are appended to the makerAssetData
```

Comment:
Same note as above - should include an offset table for reference. [Example](https://github.com/0xProject/0x-monorepo/blob/development/packages/contracts/contracts/protocol/AssetProxy/ERC20Proxy.sol#L67).
Could probably just include this once in here then reference it above.